### PR TITLE
ARROW-3778: [C++] Compile parts of test-util.h that we can once, link with unit tests

### PR DESCRIPTION
**Pull request by wesm**
This has to be compiled into a separate static library and added to the link libs for the unit tests and benchmarks

### CODE REVIEW

**Comment by wesm** [comment]
This is broken on Windows, I will have a look later today

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -733,6 +733,19 @@ endif()
 
 set(ARROW_TEST_LINK_LIBS ${ARROW_MIN_TEST_LIBS})
 
+if (ARROW_BUILD_TESTS OR ARROW_BUILD_BENCHMARKS)
+  # TODO(wesm): maybe find a better place to put shared auxiliary testing tools
+  # that depend on gtest
+  add_library(arrow_testing_aux STATIC src/arrow/test-util.cc)
```
Comment (xhochy) [inline cpp/CMakeLists.txt]: Is there a reason why this must be `STATIC`?  Maybe using `OBJECT` or `SHARED` will solve your problem on Windows, probably `OBJECT` is what you're looking for: compile once, link "statically" into the tests everywhere.

**Comment by wesm** [comment]
Argh, so the issue is more complex on Windows. The test utilities contain inline symbols from other Arrow headers, which seems to cause a conflict with the symbols in arrow.dll (multiply defined symbols). Since these are inline I don't really get why it's a problem. I don't know enough about linking on Windows to know how to fix this using the `add_library(... OBJECT)` method. 

To resolve the issue, I am now building both static and shared variants of the testing helper library and linking the corresponding library depending on whether ARROW_BUILD_SHARED is on

**Comment by pitrou** [comment]
Another possibility would be to compile `test-util` into the Arrow library (possibly inside a `arrow::test` namespace). That would increase the library size a bit but I think we can live with that.

**Comment by wesm** [comment]
We can't do that because the test code depends on gtest

**Comment by wesm** [comment]
We could of course make the compiled code not depend on gtest, but that would best be taken care of in another patch

**Comment by wesm** [comment]
Windows is a huge pain. The static CRT build is failing on Windows for reasons unknown. I will have to keep looking at this later today

**Comment by pitrou** [comment]
I'm sure that has to do with different copies of the CRT being linked in.
I think that the approach of linking `test-util` into the Arrow library may help minimize such issues.

**Comment by wesm** [comment]
I'll look at removing the gtest-depending code from the object file. That will be easier for me than messing around with linker settings

**Comment by wesm** [comment]
Started looking at this. The gtest stuff is pretty inextricable. We need to have functions like `AssertArraysEqual` that can be used with googletest that will perform appropriate assertions.

I'll keep looking at the static CRT thing for now

**Comment by wesm** [comment]
OK, the key item is here:

https://github.com/apache/arrow/blob/master/ci/appveyor-cpp-build.bat#L24

```
  @rem Since we link the CRT statically, we should also disable building
  @rem the Arrow shared library to link the tests statically, otherwise
  @rem the Arrow DLL and the tests end up using a different instance of
  @rem the CRT, which wreaks havoc.
```

But the recent change to link the unit tests dynamically is forcing ARROW_BUILD_SHARED to be on. This is a bit hacky, so I'm going to introduce an ARROW_TESTS_LINKAGE argument so we can switch to static linkage for the tests

**Comment by wesm** [comment]
Merging this

**Pull request closed without merging**